src/Query/DateParser.ts

```diff
@@ -75,12 +93,74 @@ export class DateParser {
             const unitOfTime = range === 'week' ? 'isoWeek' : (range as moment.unitOfTime.DurationConstructor);
             dateRange = [dateRange[0].startOf(unitOfTime), dateRange[1].endOf(unitOfTime)];
 
-            // Dates shall be at midnight eg 00:00
-            dateRange.forEach((d) => d.startOf('day'));
+            DateParser.setDateRangeToStartOfDay(dateRange);
 
             return dateRange;
         }
 
         return undefined;
     }
+
+    private static specificYearRegex = /[0-9]{4}/;
+    private static specificQuarterRegex = /[0-9]{4}-Q[1-4]/;
+    private static specificMonthRegex = /[0-9]{4}-[0-9]{2}/;
+    private static specificWeekRegex = /[0-9]{4}-W[0-9]{2}/;
+    private static specificYearFormat = 'YYYY';
+    private static specificQuarterFormat = 'YYYY-Q';
+    private static specificMonthFormat = 'YYYY-MM';
+    private static specificWeekFormat = 'YYYY-WW';
+
+    private static parseSpecificDateRange(input: string): [moment.Moment, moment.Moment] | undefined {
+        const yearMatch = input.match(DateParser.specificYearRegex);
+        if (yearMatch && yearMatch.length === 1 && yearMatch[0] === input) {
+            return DateParser.buildSpecificDateRange(yearMatch[0], DateParser.specificYearFormat);
+        }
```

Comment:
I am not sure that I understand the logic here.

Could there be a comment explaining the purpose of `yearMatch.length` and `yearMatch[0] === input`?

Reply:
The logic is to be sure that there is only 1 match and it is entirely the string from the input. Do you think this is an overkill?

Reply:
^ is start of the line.
$ is end.
And there would need to be a JSDoc comment saying that the caller must `trim()` whitespace from the value.

Reply:
(I mean, the current implementation requires that comment too - the regex just makes the code simpler in the implementation)

Reply:
> And there would need to be a JSDoc comment saying that the caller must `trim()` whitespace from the value.

I'm sorry but why? +)

Reply:
Because if I pass in `1927 ` it fails to parse. because the regex match is `1927` and the input is `1927 ` - so the check that `match === input` fails...

So it's a hidden part of the implementation that one cannot pass in strings with spaces surrounding the meaningful value...

Reply:
Ok so you mean that `/^[0-9]{4}$/` will accept `1927 ` and ` 1927`? In this case why not just leaving the `match === input`? Isn't this clearer than a JSDoc comment? Not against it, just curious.

Reply:
I think I have been unclear. Apologies... 😄 

The behaviour is still that this new, demo test fails:

```js
    it('should allow the range text to be padded with spaces', () => {
        testParsingDateRange('2022 ', '2022-01-01', '2022-12-31');
    });
```

The result is:

```diff

Error: expect(received).toStrictEqual(expected) // deep equality

- Expected  - 2
+ Received  + 2

  Array [
-   "2022-01-01",
-   "2022-12-31",
+   "Invalid date",
+   "Invalid date",
  ]
```

In the case of parsing a `due ...` line, this is fine, as I believe that the value is trimmed before being passed to this function.

So the behaviour is fine currently.

BUT in case anyone uses `DateParser` in a different context in new code and does not trim spaces, they will get an invalid date range for an otherwise valid specific date range string.

Options for dealing with this are:

- Shrug, say all that matters is its current use and document in JS doc for the public method `parseDateRange()` that the string must be trimmed by the caller
    - Then, we keep the code simple now and, if it is needed to parse un-trimmed values in future, it will be easy to change it to do so
- Or, now that I think of it, change the regexes to allow optional whitespace at either end of the line (and add test(s) accordingly)


```diff
-            [/^[0-9]{4}$/, 'YYYY', 'year'],
+            [/^\s*[0-9]{4}\s*$/, 'YYYY', 'year'],
```

Reply:
Thanks for the details, I get your concern now.

I think that the second solution you proposed is quite good. However this will make the spaces "slip" into the `moment(range, format)` call and I don't think it is a good idea to put responsibility on managing the spaces to an outer API and the call shall be as clean as possible so I just `trim()` the `range` after the regexp =)

Hope we can resolve this now.

In any case thanks for sharing your concern which is valid indeed, I will try to keep this approach.